### DISCUSSION TITLE
Example of boolean (Yes/No) dependent fields

### DISCUSSION THREAD

**Discussion by Svish** [jaredpalmer/formik#3064] [Q&A] [answered]
A common pattern in larger forms on our website is first a boolean "Yes"/"No" choice, and if they choose e.g. "Yes", the form extends with more fields to emphasize on that. For example "Do you have children?", and if they answer "Yes", maybe a number input of "How many under 16?".

I was hoping this would be the topic covered in your [dependent-fields](https://formik.org/docs/examples/dependent-fields) example, but unfortunately it was instead deriving a value from another field.

Can someone provide an example of how one would do a setup like this using Formik (and preferably add it to the documentation)?

---

Seeing [this issue](https://github.com/formium/formik/issues/1268) makes me a bit worried, but don't know Formik well enough to understand if the comments there are saying "this is not possible", "this is possible, but will be a pain", or "here's how to do it". I'm basically just trying to see if Formik (and Yup, probably) can do what we would need it to do for our forms (which are currently implemented in a super magic total nightmarish fashion...)

**Comment by johnrom** [comment, marked as answer]
The only issue you'll find is if you don't know those fields ahead of time (they are coming from a CMS or something).

In most cases, you know what fields need to be displayed at build time if someone chooses yes or no, in which case it's very easy to do something like this:

```tsx
return <Formik initialValues={{ hasChildren: false, howMany: 0 }}>
  <label htmlFor="hasChildrenField">Do you have children?" />
  <ToggleField name="hasChildren" id="hasChildrenField" />
  <DependentFields dependsOn="hasChildren">
    <label htmlFor="howManyField">How many?</label>
    <Field name="howMany" id="howManyField" type="number" />
  </DependentFields>
</Formik>
```

Looks simple enough, but the question is: what is Toggle Field? Let's assume you want it to be a radio input, though you might make it a Toggle. It would work the same. Once you create one component which works like this, you can use it anywhere you need it.

```tsx
import { useField } from 'formik';

export interface ToggleFieldProps {
  name: string,
}

// toggle field converts yes or no to true or false
export const ToggleField = (props: ToggleFieldProps) => {
  const [field] = useField<boolean>(props);

  // Yes = true, otherwise false
  const handleChange = useCallback((event) => field.onChange(event.target.value === "Yes"), [field.onChange]);

  // use `field` but override onChange
  return <><label>
      <Field {...field} type="radio" value="Yes" onChange={handleChange} />
      Yes
    </label>
    <label>
       <Field {...field} type="radio" value="No" onChange={handleChange} />
       No
    </label></>
}
```

Then what is DependentFields ?? Well, it's just a component that checks if values.hasChildren is true and if so, adds an active class or something.

```tsx
export const DependentFields = ({ dependsOn, children }) => {
  const [field] = useField(dependsOn);
  let className = "dependent-fields";

  if (field.value) { className += " dependent-fields--active"; }

  return <div className={className}>
    {children}
  </div>
}
```

Now the dependent fields only show if you toggle to true.

**Reply by Svish** [reply]
Right, OK. So then one would just need to do the appropriate thing when validating (e.g. use `when` with `yup`), manually remove the `howMany` field before submitting if `hasChildren` is `Yes` (if it should only be in request if `No`?

Or can `yup` or `react-query` possibly help out with that last cleanup bit too perhaps?

**Reply by Svish** [reply]
Btw., an example like that would be great to have in the docs. 👍

**Reply by johnrom** [reply]
Yup has the `when` method for this purpose. We try to avoid documenting Yup when possible, since we don't control their API. I agree we can use more examples, feel free to submit a PR!

https://stackoverflow.com/questions/49394391/conditional-validation-in-yup

**Reply by Svish** [reply]
Yeah, leaving out Yup from most documentation is probably a good idea. I just meant the ToggleField and DependentFields setup, as that's what I was actually expecting to see when I clicked on the documentation page for "Dependent fields", rather than the example that's there now which merges the values of two other fields.
